keps/sig-network/0030-nodelocal-dns-cache.md

```diff
@@ -192,7 +192,13 @@ Having the pods query the nodelocal cache introduces a single point of failure.
 
```

Comment:
I've just had a thought: `169.254.20.10` is link-local. If Node-local DNS cache relies, by default, on Pods being able to route to `169.254.20.10`  then that is a dependency on undefined / nonstandard behavior.
According to (my reading of) internet standards, `169.254.0.0/16` is only reachable for a client on the same link.

I recommend addressing that, somehow, before GA. If the decision is to go against the standard, let's record that explicitly.

Reply:
The link-local IP is attached to an interface on the host. You are right, it does not exist on each pod. Since all pod traffic does go through the host network and subject to iptables rules configured, this works. I am not aware of any CNI plugins where this would not work. From what i understand, only 127.0.00/8 traffic does not leave the pod.

Reply:
A CNI plugin will need to route traffic to 169.254.20.10, but there's no standardized mechanism to make this happen (because 169.254.20.10 is link-local, and link-local addresses aren't routeable).

Reply:
The packets just need to hit the hostNetwork since the link-local IP address exists on an interface in the node. This creates a route in the "local" routing table of the form:
"local 169.254.20.10 dev nodelocaldns proto kernel scope host src 169.254.20.10"

This is sufficient to deliver the packet to anything listening locally on that address. Isn't that true?

Also, if the k8s cluster is using kube-proxy in iptables mode, the pods can use either the link-local IP or the kube-dns service IP to access node-local-dns.

Reply:
> `local 169.254.20.10 dev nodelocaldns proto kernel scope host src 169.254.20.10`

I'm really not keen on setting that in stone. Allocating from the link-local range like this and expecting routing to work, even within only one host, just doesn't seem like an approach that can scale.

It's different when a cloud provider does it at the network layer, because there's no concept of those packets being _routed_. There's a (virtual) link and it turns out that when the instance boots something else is defending `169.254.169.254` or whatever. That works fine.


Ideally, I can set up a Kubernetes cluster that puts all Pods on the same link-local network, and that's a valid supported configuration (provided I find a CNI that makes it happen). I realize that `169.254.0.0/16` looks like a convenient range to borrow an address from, and I can see why it's _often_ going to be a pragmatic choice.

To be clear: my concern is about making & documenting `169.254.20.10` as a _default_ value. If you have to provide an IPv4 address at deploy time and the documentation suggests picking an address inside `169.254.0.0/16` then that isn't a problem and doesn't set any problematic precedent.

Reply:
ok, if I understand correctly, we want to tweak the documentation to say:
"Selecting a listen ip for NodeLocal DNSCache from 169.254.20.10/16 is recommended, in our examples we will use 169.254.20.10". Is that correct? I have added this to the website doc PR - https://github.com/kubernetes/website/pull/18716

We can continue the discussion there. From the GA graduation perspective, are the changes here ok to implement?

Reply:
> Selecting a listen ip for NodeLocal DNSCache from 169.254.20.10/16 is recommended, in our examples we will use 169.254.20.10

(Retrospectively) I think you've taken an appropriate and pragmatic tack. Thanks.